server/migrations/20210429022658-fix-substrate-address.js

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+const { checkAddress, decodeAddress, encodeAddress } = require('@polkadot/util-crypto');
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    const addresses = await queryInterface.sequelize.query('SELECT id, address, chain FROM "Addresses";');
+    const chains = await queryInterface.sequelize.query('SELECT id, network, base, ss58_prefix FROM "Chains";');
+    const promises = [];
+    const removeAddressIds = [];
+
+    for (const addr of addresses[0]) {
+      const { id, address, chain } = addr;
+      const ch = chains[0].find((_) => _.network === chain);
+      if (ch && ch.base === 'substrate' && !!ch.ss58_prefix) {
+        let decodedAddress;
+        try {
+          decodedAddress = decodeAddress(address);
+        } catch (e) {
+          throw new Error('failed to decode address');
+        }
+
+        // check if it is valid with the current prefix & reencode if needed
+        const [valid] = checkAddress(address, ch.ss58_prefix);
+        if (!valid) {
+          try {
+            const encoded = encodeAddress(decodedAddress, ch.ss58_prefix);
+            const duplicated = await queryInterface.sequelize.query(`SELECT id FROM "Addresses" WHERE address='${encoded}' AND chain='${chain}';`);
+            if (duplicated[0] && duplicated[0].length) {
+              removeAddressIds.push(id);
```

Comment:
So we're resolving conflicts here by removing the misencoded one? What's the reason for that approach?

Reply:
The current migration script's purpose is to format the substrate address correctly based on its chain's ss58_prefix.
For example, right now there is a wrongly formatted `Darwinia` address in `Address` table.
m7yJvcrYhKq7MmrEHLusCx1nRAz2Db7HF76ZHkHUC5kG9yP
This is a wrongly formatted address and this migration script will fix this as `2xxxxxx`.
But for some reason, there is already an address with `2xxxxxx` in Darwinia. So it causes the unique constraint error.
In this case, I am going to remove the wrongly formatted address and its related records in other tables.

Reply:
Right, I understand, I'm asking for the reason to remove the wrongly formatted address rather than selecting between the two based on some other criteria, like whichever was created first, or which is more active, or something like that.

Reply:
Yes. That makes sense. But this issue is being happened since we merge the `unifying chainbase` stuff. So it was not far and I think, the wrongly formatted addresses are less active and created later.

Reply:
Sounds good -- just wanted to know the reasoning.